streams/runners/storm/sampling/src/main/java/com/hortonworks/streamline/streams/sampling/service/storm/StormTopologySamplingService.java

```diff
@@ -0,0 +1,257 @@
+    }
+
+    @Override
+    public void init(Map<String, Object> conf) {
+        String stormApiRootUrl = null;
+        Subject subject = null;
+        if (conf != null) {
+            stormApiRootUrl = (String) conf.get(TopologyLayoutConstants.STORM_API_ROOT_URL_KEY);
+            subject = (Subject) conf.get(TopologyLayoutConstants.SUBJECT_OBJECT);
+        }
+        Client restClient = ClientBuilder.newClient(new ClientConfig());
+        this.client = new StormRestAPIClient(restClient, stormApiRootUrl, subject);
+    }
+
+    @Override
+    public boolean enableSampling(Topology topology, int pct, String asUser) {
+        String topologyId = StormTopologyUtil.findStormTopologyId(client, topology.getId(), asUser);
+        return client.enableSampling(topologyId, pct, asUser);
+    }
+
+    @Override
+    public boolean enableSampling(Topology topology, TopologyComponent component, int pct, String asUser) {
+        String topologyId = StormTopologyUtil.findStormTopologyId(client, topology.getId(), asUser);
+        return client.enableSampling(topologyId, component.getId() + "-" + component.getName(), pct, asUser);
+    }
+
+    @Override
+    public boolean disableSampling(Topology topology, String asUser) {
+        String topologyId = StormTopologyUtil.findStormTopologyId(client, topology.getId(), asUser);
+        return client.disableSampling(topologyId, asUser);
+    }
+
+    @Override
+    public boolean disableSampling(Topology topology, TopologyComponent component, String asUser) {
+        String topologyId = StormTopologyUtil.findStormTopologyId(client, topology.getId(), asUser);
+        return client.disableSampling(topologyId, component.getId() + "-" + component.getName(), asUser);
+    }
+
+    @Override
+    public SamplingStatus getSamplingStatus(Topology topology, String asUser) {
+        String topologyId = StormTopologyUtil.findStormTopologyId(client, topology.getId(), asUser);
+        return buildSamplingStatus(client.getSamplingStatus(topologyId, asUser));
+    }
+
+    @Override
+    public SamplingStatus getSamplingStatus(Topology topology, TopologyComponent component, String asUser) {
+        String topologyId = StormTopologyUtil.findStormTopologyId(client, topology.getId(), asUser);
+        return buildSamplingStatus(client.getSamplingStatus(topologyId, asUser));
+    }
+
+    @Override
+    public SampledEvents getSampledEvents(Topology topology, TopologyComponent component, EventQueryParams qps, String asUser) {
+        List<SampledEvent> events = new ArrayList<>(qps.count());
+        String topologyId = StormTopologyUtil.findStormTopologyId(client, topology.getId(), asUser);
+        String componentId = component.getId() + "-" + component.getName();
+        Long cur;
+        if (qps.desc()) {
+            cur = qps.start();
+        } else {
+            cur = qps.start() == null ? Long.valueOf(0) : qps.start();
+        }
+        Long nextOffset;
+        Long prevOffset = null;
+        do {
```

Comment:
Looks like this implementation may not be able to handle corner case: the event being placed to two pages, in other words, the line is exposed partially because we are splitting page by bytes. That's not only the event is hidden, but also the parsing logic for line could be broken.

Reply:
yes theres a chance that the partial events at the end can be missed based on the offset and its really difficult to avoid this while handling byte offsets. I have added checks to skip such partial lines. Similar behavior can happen while viewing the event logs via Storm's log viewer as well. 

When we enhance the sampling with event co-relation and implement at SAM level, we would rather store the events in a DB which will not have this limitation and would be easy to query.